cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ProgramScheduleService.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.scheduler;
+
+import co.cask.cdap.api.app.ApplicationSpecification;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.BadRequestException;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramScheduleStatus;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.ScheduleId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.security.authorization.AuthorizationUtil;
+import co.cask.cdap.security.spi.authentication.AuthenticationContext;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Used to interact with the Scheduler.
```

Comment:
This class is actually partial. Currently the logic that involves both program and schedule are scattered. E.g. in the ProgramLifecycleHttpHandler, it has the logic to add schedule. It's probably better to either move those logic here, or not to introduce this class now, but just have these two methods go into `ProgramLifecycleHttpHandler`, since they are only called from there.

Reply:
I'll move logic from the handler into this class.